File: Java/src/com/daily/algothrim/sort/QuickSort.java
package com.daily.algothrim.sort;

/**
 * 快速排序
 * 不稳定排序
 * 原地排序
 */
public class QuickSort {

    public static void main(String[] args) {
        int[] a = new int[]{10, 1, 4, 2, 5, 643, 654, 234, 2};
        new QuickSort().sort(a, 0, a.length - 1);
        for (int value : a) {
            System.out.println(value);
        }
    }

    /**
     * 时间复制度O(nlog n); 最好时间复杂度O(nlog n); 最坏时间复杂度O(n2)
     * 空间复杂度O(1)
     */
    private void sort(int[] a, int start, int end) {
        if (start < end) {
            int mid = quick(a, start, end);
            sort(a, start, mid - 1);
            sort(a, mid + 1, end);
        }
    }

    private int quick(int[] a, int start, int end) {
        int i = start;
        int j = start;
        int pivot = a[end];

        for (; j < end; j++) {
            if (a[j] < pivot) {
                int temp = a[i];
                a[i] = a[j];
                a[j] = temp;
                i++;
            }
        }

        int temp = a[i];
        a[i] = pivot;
        a[end] = temp;

        return i;
    }
}

File: kotlin/src/com/daily/algothrim/sort/QuickSort.kt
package com.daily.algothrim.sort

/**
 * 快速排序
 * 不稳定排序
 * 原地排序
 */
class QuickSort {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val a = intArrayOf(10, 1, 4, 2, 5, 643, 654, 234, 2)
            QuickSort().quick(a, 0, a.size - 1)
            a.forEach {
                println(it)
            }
        }
    }

    /**
     * 时间复制度O(nlog n); 最好时间复杂度O(nlog n); 最坏时间复杂度O(n2)
     * 空间复杂度O(1)
     */
    private fun quick(a: IntArray, start: Int, end: Int) {
        if (start < end) {
            // 找到分区位置
            val mid = sort(a, start, end)
            // 分区
            quick(a, start, mid - 1)
            quick(a, mid + 1, end)
        }
    }

    private fun sort(a: IntArray, start: Int, end: Int): Int {
        val pivot = a[end]
        var i = start
        var j = start

        while (j < end) {
            if (a[j] < pivot) {
                val temp = a[i]
                a[i] = a[j]
                a[j] = temp
                i++
            }
            j++
        }

        val temp = a[i]
        a[i] = pivot
        a[end] = temp

        return i
    }
}